### ISSUE TITLE
Clean up of fre make code

### ISSUE THREAD

**Issue by singhd789** [NOAA-GFDL/fre-cli#190]
**Is your feature request related to a problem? Please describe.**
There are multiple places where code could be broken up or cleaned to follow the flow of fre make easier. 

1) Very few instances where checking equality to a boolean object is preferred over using the boolean object value itself explcitly- `if self.repro` is sufficient for example

**Describe the solution you'd like**
**As per Ian's suggestions**
1) When checking equality to a boolean object, there are few instances where this is preferred over using the boolean object value itself. 
Instance: https://github.com/NOAA-GFDL/fre-cli/blob/5736653b0dcd68ec274d2f4754c129b4736a832a/fre/make/gfdlfremake/targetfre.py#L50
- [ ] `if self.repro` is sufficient

2) Line seems dense: https://github.com/NOAA-GFDL/fre-cli/blob/5736653b0dcd68ec274d2f4754c129b4736a832a/fre/make/gfdlfremake/makefilefre.py#L108
- [ ] try to break this up

3) Class logic needs updating/refactoring. If we want to go in the direction of OOP, it would make much more sense to map the yaml config's values to explicit fields of the compileYaml class. e.g., instead of what's currently shown on [L53-L62](https://github.com/NOAA-GFDL/fre-cli/blob/5736653b0dcd68ec274d2f4754c129b4736a832a/fre/make/gfdlfremake/yamlfre.py#L58)
- [ ] Map config values to explicit fields. Could look like this:
```
        # compile information from the combined yaml
        self.yaml = compileinfo

        ## Check the yaml for required things
        ## Check for required experiment name
        try:
            self.experiment=compileinfo['experiment']
        except:
            raise ValueError("You must set an experiment name to compile")
```

**Comment by ilaflott** [comment]
from https://github.com/NOAA-GFDL/fre-cli/pull/304#discussion_r1901742948:

For dev history (a.k.a. posterity), print statement itself is not really the repeat that i'm noticing, though it does seem duplicated and we need to replace print statements with `logging`. but that's a different thing.

the similarity can be most easily seen across these links:
[`create_checkout_script`](https://github.com/NOAA-GFDL/fre-cli/blob/e699d0807946931e6b5dd213293d4f49d28e2bb2/fre/make/create_checkout_script.py#L60-L99)
[`create_compile_script`](https://github.com/NOAA-GFDL/fre-cli/blob/e699d0807946931e6b5dd213293d4f49d28e2bb2/fre/make/create_compile_script.py#L51-L88)
[`create_docker_script`](https://github.com/NOAA-GFDL/fre-cli/blob/e699d0807946931e6b5dd213293d4f49d28e2bb2/fre/make/create_docker_script.py#L37-L79)
[`create_makefile_script`](https://github.com/NOAA-GFDL/fre-cli/blob/e699d0807946931e6b5dd213293d4f49d28e2bb2/fre/make/create_makefile_script.py#L34-L77)

basically, it looks like we need an abstract base class/method, `script_maker`, and have `makefile_script_maker`, `compile_script_maker`, `checkout_script_maker`, and `docker_script_maker` all inherit from that class or build on the base function.

To a lesser extent, `run_fremake_script` also has a very similar structure, but it's name implies a more distinct purpose, so i'm leaving it out thinking it wouldn't generalize to the other four simply/trivially.

**Comment by ilaflott** [comment]
use `return` and `raise` instead of [`sys.exit`](https://github.com/search?q=repo%3ANOAA-GFDL%2Ffre-cli%20sys.exit&type=code), which can ruin backtracing.

**Comment by ilaflott** [comment]
certain tests in `fre/make/tests` are expected to fail in certain contexts (e.g. workstation). e.g. rig the `pytest` logic to skip the test if it's not the CI image for `fre/make/tests/compilation/test_run_fremake_builds.py`.

**Comment by ilaflott** [comment]
certain tests in `fre/tests` will create `fre/tests/fremake_out`, which contains a checkout of [`fre-workflows`](https://github.com/noaa-gfdl/fre-workflows.git). nothing inherently wrong with that.

but because `fre-workflows` has python, if one wants to run `pytest` twice to test another edit, the `pytests` in `fre-workflows` are getting picked up the second time around, causing a failure and demanding manual removal of the files before re-running. not fun for devlopment! the tests should clean up after themselves.
